File: ThreeSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 1 && nums[i] == nums[i - 1]) {
                continue;
            }
            int left = i + 1;
            int right = nums.length - 1;
            while (left <= right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    List<Integer> cur = new ArrayList<>();
                    cur.add(nums[i]);
                    cur.add(nums[left]);
                    cur.add(nums[right]);
                    res.add(cur);
                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    left++;
                    right--;
                } else if (sum > 0) {
                    right--;
                } else {
                    left++;
                }
            }
        }
        return res;
    }
}

File: Threesum.py
class Solution:
    def threesum(self,nums : List[int]) -> List[List[int]]:
        nums.sort()
        triplet = set()
        for i in range(len(nums)-2):
            firstNum = nums[i]
            j=i+1
            k=len(nums)-1
            while j<k: 
                secondNum = nums[j]
                thirdNum = nums[k]

                potentialSum = firstNum+secondNum+thirdNum
                if potentialSum > 0:
                    k-=1
                elif potentialSum <0 :
                    j+=1
                else:
                    triplet.add([firstNum,secondNum,thirdNum])
                    j+=1
                    k-=1
        return triplet

